File: python/examples/basic_usage.py
"""
Basic usage example for Omne Python SDK

This example demonstrates fundamental operations:
- Connecting to Omne network
- Wallet creation and management
- Balance checking
- Simple transactions
"""

import asyncio
from decimal import Decimal
from omne_sdk import OmneClient, Wallet


async def basic_usage_example():
    """Demonstrate basic Omne SDK usage"""
    
    print("🚀 Omne Python SDK - Basic Usage Example")
    print("=" * 50)
    
    # Connect to Omne network
    client = OmneClient("http://localhost:8545")
    
    try:
        # Get network information
        print("\n📡 Network Information:")
        network_info = await client.get_network_info()
        print(f"  Network: {network_info.network_name}")
        print(f"  Chain ID: {network_info.chain_id}")
        print(f"  API Version: {network_info.api_version}")
        print(f"  ORC-20 Enabled: {network_info.features.get('orc20Enabled', False)}")
        print(f"  OON Enabled: {network_info.features.get('oonEnabled', False)}")
        
        # Generate new wallet
        print("\n🔐 Wallet Generation:")
        wallet = Wallet.generate()
        print(f"  Address: {wallet.address}")
        print(f"  Mnemonic: {wallet.mnemonic}")
        
        # Get multiple accounts from wallet
        print("\n👥 Account Derivation:")
        for i in range(3):
            account = wallet.get_account(i)
            print(f"  Account {i}: {account.address}")
        
        # Check balance (will be 0 for new address)
        print("\n💰 Balance Check:")
        balance = await client.get_balance(wallet.address)
        print(f"  Balance: {balance.omc} OMC ({balance.quar:,} quar)")
        print(f"  Last Updated: Block #{balance.last_updated}")
        
        # Demonstrate utility functions
        print("\n🔧 Utility Functions:")
        test_amounts = [1.0, 0.5, 0.001, 10.5]
        for amount in test_amounts:
            quar = client.to_quar(amount)
            back_to_omc = client.from_quar(quar)
            print(f"  {amount} OMC = {quar:,} quar = {back_to_omc} OMC")
        
        # Address validation
        print("\n✅ Address Validation:")
        test_addresses = [
            wallet.address,
            "0x742d35cc4bf688aee6f7c3c3a6b1c98aaee5e84e",
            "invalid_address",
            "0x123"  # Too short
        ]
        
        for addr in test_addresses:
            is_valid = client.is_valid_address(addr)
            status = "✅ Valid" if is_valid else "❌ Invalid"
            print(f"  {addr}: {status}")
        
        # Get current gas price
        print("\n⛽ Gas Information:")
        gas_price = await client.get_gas_price()
        print(f"  Current gas price: {gas_price:,} quar per gas")
        print(f"  Standard transfer cost: ~{21000 * gas_price:,} quar")
        print(f"  Standard transfer cost: ~{client.from_quar(21000 * gas_price)} OMC")
        
        # Get latest block
        print("\n📦 Latest Block:")
        latest_block = await client.get_block("latest")
        print(f"  Block Number: #{latest_block.number}")
        print(f"  Block Hash: {latest_block.hash}")
        print(f"  Timestamp: {latest_block.timestamp}")
        print(f"  Layer: {latest_block.layer}")
        print(f"  Transactions: {latest_block.transaction_count}")
        
        print("\n✅ Basic usage example completed successfully!")
        
    except Exception as e:
        print(f"\n❌ Error: {e}")
        print("Make sure Omne node is running on localhost:8545")
        
    finally:
        await client.close()


if __name__ == "__main__":
    asyncio.run(basic_usage_example())

File: typescript/examples/basic-usage.ts
/// <reference types="node" />

/**
 * Basic usage example for Omne TypeScript SDK
 * 
 * Demonstrates core functionality including wallet creation,
 * transactions, and quar-precision calculations.
 */

import { 
  Wallet, 
  toQuar, 
  fromQuar, 
  formatBalance,
  createClient 
} from '../src/index';

async function basicUsageExample() {
  console.log('🚀 Omne TypeScript SDK - Basic Usage Example');
  console.log('=============================================\n');

  // 1. Create client connection
  console.log('📡 Connecting to Omne network...');
  const rpcUrl = process.env.OMNE_RPC_URL || 'http://127.0.0.1:8545';
  const client = createClient(rpcUrl);
  
  try {
    await client.connect();
    console.log(`✅ Connected to Omne network at ${rpcUrl}\n`);
  } catch (error) {
    console.log(`⚠️ Connection failed for ${rpcUrl}`);
    console.log('📝 Falling back to offline demonstrations...\n');
  }

  // 2. Wallet operations
  console.log('💰 Wallet Operations');
  console.log('--------------------');

  // Generate new wallet
  const wallet = Wallet.generate();
  console.log(`Mnemonic: ${wallet.getMnemonic()}`);
  console.log(`Address: ${wallet.address}\n`);

  // Derive multiple accounts
  const accounts = wallet.getAccounts(3);
  accounts.forEach((account, index) => {
    console.log(`Account ${index}: ${account.address}`);
  });
  console.log();

  // 3. Quar precision calculations
  console.log('🔢 Quar Precision Demonstrations');
  console.log('--------------------------------');

  const amounts = [1.0, 0.5, 0.001, 10.5, 0.0001];
  amounts.forEach(amount => {
    const quar = toQuar(amount);
    const omc = fromQuar(quar);
    console.log(`${amount} OMC = ${quar} quar = ${omc.toString()} OMC`);
  });
  console.log();

  // 4. Commerce scenarios
  console.log('🏪 Commerce Transaction Examples');
  console.log('--------------------------------');

  const commerceScenarios = [
    { name: 'Coffee', price: 0.003 },
    { name: 'Digital download', price: 0.15 },
    { name: 'Subscription', price: 9.99 },
    { name: 'Microtransaction', price: 0.0001 }
  ];

  commerceScenarios.forEach(scenario => {
    const valueQuar = toQuar(scenario.price);
    const gasQuar = 21000 * 1000; // 21k gas * 1000 quar/gas
    const totalQuar = BigInt(valueQuar) + BigInt(gasQuar);
    const totalOMC = fromQuar(totalQuar.toString());
    
    console.log(`${scenario.name}: ${scenario.price} OMC + gas = ${totalOMC.toString()} OMC total`);
  });
  console.log();

  // 5. Transaction creation (offline)
  console.log('📝 Transaction Creation');
  console.log('----------------------');

  const account = wallet.getAccount(0);
  const transaction = {
    from: account.address,
    to: '0x742d35cc4bf688aee6f7c3c3a6b1c98aaee5e84e',
    value: toQuar(0.001), // 0.001 OMC
    gasLimit: 21000,
    gasPrice: '1000', // 1000 quar per gas
    nonce: 0,
    priority: 'commerce' as const
  };

  try {
    const signedTx = account.signTransaction(transaction);
    console.log(`Transaction hash: ${signedTx.hash || 'Generated offline'}`);
    console.log(`From: ${signedTx.from}`);
    console.log(`To: ${signedTx.to}`);
    console.log(`Value: ${formatBalance(signedTx.value)}`);
    console.log(`Priority: ${signedTx.priority}`);
    console.log(`Signature: ${signedTx.signature ? signedTx.signature.substring(0, 20) + '...' : 'N/A'}`);
  } catch (error) {
    console.log('⚠️ Transaction signing failed (expected without crypto dependencies)');
    console.log('📋 Transaction details:');
    console.log(`From: ${transaction.from}`);
    console.log(`To: ${transaction.to}`);
    console.log(`Value: ${formatBalance(transaction.value)}`);
  }
  console.log();

  // 6. SDK capabilities
  console.log('✨ SDK Feature Overview');
  console.log('----------------------');
  console.log('🔗 Blockchain Integration: Full dual-layer PoVERA support');
  console.log('💰 Microscopic Fees: Quar-precision gas calculations');
  console.log('🪙 ORC-20 Tokens: Deploy and manage application tokens');
  console.log('🤖 AI/ML Services: Submit computational jobs to OON');
  console.log('🔐 Wallet Management: BIP39 HD wallets with signing');
  console.log('⚡ Async/Await: Modern JavaScript async patterns');
  console.log('🔒 Type Safety: Full TypeScript type definitions');
  console.log();

  console.log('🎉 Omne TypeScript SDK demonstration complete!');
  
  // Cleanup
  await client.disconnect();
}

// Run example if called directly
if (require.main === module) {
  basicUsageExample().catch(console.error);
}

export { basicUsageExample };
